sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/EFOShardSubscribersPool.java

```diff
@@ -196,33 +200,45 @@ public void start(Iterable<ShardCheckpoint> checkpoints) {
    */
   @Nullable
   KinesisRecord getNextRecord() throws IOException {
-    if (current == null && eventQueue.isEmpty()) {
-      if (subscriptionError == null) {
-        return null;
-      } else {
-        throw new IOException(subscriptionError);
+    while (true) {
```

Comment:
This must be wrapped in a while loop. We only return `null` if the queue is empty!

Reply:
`while` gave struggles with tests. Then I came to think:
- waiting indefinite time felt like less control
- for me, having an indefinite loop was more cognitive overhead and difficult to comprehend and not to forget all the cases where we must keep inside it and where we must exit it, how the state was updated, etc.
- if we have `eventQueue.isEmpty()` to be false, what's the chance `poll()` yields nothing and why we re-poll-ing?
- without the loop it seemed to work just fine with Flink runner

Having all that in mind, I think, I got simpler version of `getNextRecord()`

Let's discuss this more, I might be missing something important here.